Added try-catch to ScriptEngine.cs

dynms1337 opened the pull request:
Added try-catch to display helpful log for investigation when dll files could not be reloaded.

- Before
```
[Error  : Unity Log] ReflectionTypeLoadException: The classes in the module cannot be loaded.
Stack trace:
System.Reflection.Assembly.GetTypes ()
ScriptEngine.ScriptEngine.LoadDLL (System.String path, UnityEngine.GameObject obj)
ScriptEngine.ScriptEngine.ReloadPlugins ()
ScriptEngine.ScriptEngine.Update ()
```

- After
```
[Error  :Script Engine] Error While loading C:\(path)\BepInEx\scripts\BepInEx.ScriptSelector.dll
-- Successfully Read Types --
ScriptSelector.MonoCompiler

ScriptSelector.ScriptInfo
ScriptSelector.ScriptInfo+<>c
ScriptSelector.ScriptSelector
ScriptSelector.ScriptSelector+<>c__DisplayClass10_0
ScriptSelector.ScriptSelector+<>c
ScriptSelector.LoggerTextWriter
ScriptSelector.Utilities

-- LoaderExceptions --
System.TypeLoadException: Could not load type '<>c__DisplayClass2_0' from assembly 'BepInEx.ScriptSelector-637339073995094410, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.

-- StackTrace --
  at (wrapper managed-to-native) System.Reflection.Assembly:GetTypes (bool)
  at System.Reflection.Assembly.GetTypes () [0x00000] in <filename unknown>:0
  at ScriptEngine.ScriptEngine.LoadDLL (System.String path, UnityEngine.GameObject obj) [0x00000] in <filename unknown>:0
```

dynms1337 committed:
Added try-catch to ScriptEngine.cs

Added try-catch to display helpful information for investigation when dll files could not be reloaded

ghorsington commented on src/ScriptEngine/ScriptEngine.cs:
```diff
@@ -76,22 +76,33 @@ void LoadDLL(string path, GameObject obj)
                     }
+                    catch (ReflectionTypeLoadException e)
+                    {
+                        StringBuilder strTypes = new StringBuilder();
```
Prefer using `var` with non-primitive types.

ghorsington commented on src/ScriptEngine/ScriptEngine.cs:
```diff
@@ -76,22 +76,33 @@ void LoadDLL(string path, GameObject obj)
+                    catch (ReflectionTypeLoadException e)
+                    {
+                        StringBuilder strTypes = new StringBuilder();
+                        foreach (var t in e.Types) { strTypes.Append(t + "\r\n"); }
```
The curly brackets are not needed for single-line `foreach` blocks. In addition, put the single line on a newline likes so:

```cs
foreach (var t in e.Types)
    strTypes.Append(t + "\r\n");
```

ghorsington commented on src/ScriptEngine/ScriptEngine.cs:
```diff
@@ -76,22 +76,33 @@ void LoadDLL(string path, GameObject obj)
+                    {
+                        StringBuilder strTypes = new StringBuilder();
+                        foreach (var t in e.Types) { strTypes.Append(t + "\r\n"); }
+                        StringBuilder strExceptions = new StringBuilder();
```
Prefer using `var` with non-primitive types.

ghorsington commented on src/ScriptEngine/ScriptEngine.cs:
```diff
@@ -76,22 +76,33 @@ void LoadDLL(string path, GameObject obj)
+                        StringBuilder strTypes = new StringBuilder();
+                        foreach (var t in e.Types) { strTypes.Append(t + "\r\n"); }
+                        StringBuilder strExceptions = new StringBuilder();
+                        foreach (var l in e.LoaderExceptions) { strExceptions.Append(l + "\r\n"); }
```
The curly brackets are not needed for single-line `foreach` blocks. In addition, put the single line on a newline.

ghorsington commented on src/ScriptEngine/ScriptEngine.cs:
```diff
@@ -76,22 +76,33 @@ void LoadDLL(string path, GameObject obj)
+                        foreach (var t in e.Types) { strTypes.Append(t + "\r\n"); }
+                        StringBuilder strExceptions = new StringBuilder();
+                        foreach (var l in e.LoaderExceptions) { strExceptions.Append(l + "\r\n"); }
+                        Logger.LogError($"Error While loading {path} \r\n -- Types --\r\n{strTypes}\r\n-- LoaderExceptions --\r\n{strExceptions}\r\n -- StackTrace --\r\n{e.StackTrace}");
```
Hmmm, this looks a tad hard to read. How about making only one `StringBuilder` and build the whole error message with it? That way you can make use of `AppendLine` which will remove the need for the vast number of manual newlines.

So in the end it would only be

```cs
Logger.LogError(sbMessage.ToString());
```

I don't think you need two separate `StringBuilder`s either, you can just make do with one of them.

ghorsington reviewed (changes requested):
Thank you for the PR!

I've added a few comments regarding matching code style to one we use and one bigger comment about simplifying printing of the message.

dynms1337 committed:
Readability improvements of ScriptEngine.cs

- Use `var`
- Removed curly brackets from foreach blocks.
- Build the whole error message with one StringBuilder.

dynms1337 commented:
Thank you for Reviewing.
I've taken all the changes and committed.

ghorsington commented on src/ScriptEngine/ScriptEngine.cs:
```diff
@@ -76,22 +76,39 @@ void LoadDLL(string path, GameObject obj)
+                    catch (ReflectionTypeLoadException e)
+                    {
+                        var sbMessage = new StringBuilder();
+                        sbMessage.AppendLine($"Error While loading {path}");
```
I don't think you need to capitalize "While"

ghorsington commented on src/ScriptEngine/ScriptEngine.cs:
```diff
@@ -76,22 +76,39 @@ void LoadDLL(string path, GameObject obj)
+                        sbMessage.AppendLine($"Error While loading {path}");
+                        sbMessage.AppendLine("-- Successfully Read Types --");
+                        foreach (var t in e.Types)
+                            if (t != null) sbMessage.AppendLine(t.ToString());
```
Is printing all loaded types warranted? My main concern is that if the assembly has a dozen of them, printing all loaded types seems like it could bloat the console.

dynms1337 replied:
I thought about it, and decided to follow.
Because I wanted to output much information as possible, I added `foreach (var t in e.Types)`. After all, I didn't use it. Because the elimination method was tough work.
I think We should avoid the risk of bloating the console.

ghorsington commented on src/ScriptEngine/ScriptEngine.cs:
```diff
@@ -76,22 +76,39 @@ void LoadDLL(string path, GameObject obj)
+                            if (t != null) sbMessage.AppendLine(t.ToString());
+                        sbMessage.AppendLine("\r\n-- LoaderExceptions --");
+                        foreach (var l in e.LoaderExceptions)
+                            if (l != null) sbMessage.AppendLine(l.ToString());
```
Is there a need for a null check here? I know that `e.Types` can contain null entries, but I'm fairly certain `e.LoaderExceptions` has no nulls. 

For example, in [`TypeLoader.TypeLoadExceptionToString`](https://github.com/BepInEx/BepInEx/blob/master/BepInEx/Bootstrap/TypeLoader.cs#L232-L259) that we use for printing `ReflectionTypeLoadException` in BepInEx itself we don't have such a check.

dynms1337 replied:
I'll follow. I wondered if `e.LoaderExceptions` has nulls, but chose safe way without enough survey ^^;

ghorsington reviewed (commented):
Thanks, looks mainly good to me!

I added a few small suggestions, feel free to comment on those and potentially change them with if you want.

If you don't think that these warrant any changes, do comment so I know you won't be changing any of this. Other than that, I'll merge once these are resolved in some way.

dynms1337 committed:
Modified error log format of ScriptEngine.cs

- Lowercase "While" in error messages
- Removed null check of `e.LoaderExceptions`
- Removed printing all loaded types.

dynms1337 commented:
Thanks for comments.
Finally, I've taken all suggestions, and the error log looks like below.
```
[Error  :Script Engine] Error while loading C:\(path)\BepInEx\scripts\BepInEx.ScriptSelector.dll

-- LoaderExceptions --
System.TypeLoadException: Could not load type '<>c__DisplayClass2_0' from assembly 'BepInEx.ScriptSelector-637340811771365543, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.

-- StackTrace --
  at (wrapper managed-to-native) System.Reflection.Assembly:GetTypes (bool)
  at System.Reflection.Assembly.GetTypes () [0x00000] in <filename unknown>:0
  at ScriptEngine.ScriptEngine.LoadDLL (System.String path, UnityEngine.GameObject obj) [0x00000] in <filename unknown>:0
```

ghorsington commented:
Looks good to me!

Thank you for all the work!
